File: Typescript/typescript_ast/src/csv_output.ts
import fs from "fs";
import path from "path";
import * as Config from "../config";
import * as MethodInfo from "./method_info";

// ===================================================
// csv_output.ts
// ソースファイルレベルCSV・メソッドレベルCSVの2種類のみを出力する。
// JSON出力は行わない（json_output.tsは廃止）。
//
// エスケープ処理は、前回tsjs_ast/src/csv_output.tsで修正済みの
// 「ダブルクォートの全置換版」をそのまま踏襲する。
// map()内で使うため、ここでは同期関数として実装する
// （元のescapeCsvはasyncだったが、for文でのawait前提の設計であり、
// map()に直接渡すとPromiseの配列になってしまうため、同期関数に変更した）。
// ===================================================

export function writeSourceFileCsv(rows: MethodInfo.SourceFileRow[]) {
  const header = MethodInfo.SOURCE_FILE_ROW_HEADER;
  const headerLine = header.map(escapeCsv).join(",");

  const lines = rows.map((r) =>
    [
      escapeCsv(r.appName),
      escapeCsv(r.fileName),
      escapeCsv(r.directoryPath),
      escapeCsv(r.className),
      escapeCsv(r.importList),
      escapeCsv(String(r.lineCount)),
      escapeCsv(String(r.methodCount)),
      escapeCsv(String(r.variableCount)),
      escapeCsv(String(r.constantCount)),
    ].join(",")
  );

  const content = `${headerLine}\n${lines.join("\n")}`;
  writeFile(Config.SOURCE_FILE_CSV_FILENAME, content);
}

export function writeMethodCsv(rows: MethodInfo.MethodRow[]) {
  const header = MethodInfo.buildMethodRowHeader();
  const headerLine = header.map(escapeCsv).join(",");

  const lines = rows.map((r) => {
    const cols: string[] = [
      escapeCsv(r.filePath),
      escapeCsv(r.className),
      escapeCsv(r.methodName),
    ];
    for (let i = 0; i < MethodInfo.MAX_NEST_DEPTH; i++) {
      cols.push(escapeCsv(String(r.processCoords[i] ?? 0)));
    }
    cols.push(escapeCsv(r.processContent));
    cols.push(escapeCsv(r.role));
    cols.push(escapeCsv(r.returnType));
    cols.push(escapeCsv(r.methodType));
    cols.push(escapeCsv(r.accessModifier));
    for (let i = 0; i < MethodInfo.MAX_ARG_COLUMNS; i++) {
      cols.push(escapeCsv(r.args[i] ?? MethodInfo.HYPHEN));
    }
    return cols.join(",");
  });

  const content = `${headerLine}\n${lines.join("\n")}`;
  writeFile(Config.METHOD_CSV_FILENAME, content);
}

// 【新規】フィールドレベルCSV（3つ目のCSV）の出力
export function writeFieldCsv(rows: MethodInfo.FieldRow[]) {
  const headerLine = MethodInfo.FIELD_ROW_HEADER.map(escapeCsv).join(",");

  const lines = rows.map((r) =>
    [
      escapeCsv(r.filePath),
      escapeCsv(r.className),
      escapeCsv(r.methodName ?? ""),
      escapeCsv(r.fieldKind),
      escapeCsv(r.fieldName),
      escapeCsv(r.fieldType),
      escapeCsv(r.isFinal ? "1" : "0"),
      escapeCsv(String(r.validationMin)),
      escapeCsv(String(r.validationMax)),
      escapeCsv(String(r.nullable)),
      escapeCsv(r.rawAnnotations),
    ].join(",")
  );

  const content = `${headerLine}\n${lines.join("\n")}`;
  writeFile(Config.FIELD_CSV_FILENAME, content);
}

function writeFile(filename: string, content: string) {
  if (!fs.existsSync(Config.DEFAULT_OUTPUT_DIR)) {
    fs.mkdirSync(Config.DEFAULT_OUTPUT_DIR, { recursive: true });
  }
  fs.writeFileSync(path.join(Config.DEFAULT_OUTPUT_DIR, filename), content, Config.ENCODING as BufferEncoding);
  console.log(`${filename}: ${content.split("\n").length - 1}行`);
}

// escapeCsv: ダブルクォートは全置換（前回修正済みの挙動を踏襲）
export function escapeCsv(jsonValue: any): string {
  const stringValue = String(jsonValue);
  if (/[",\n\r]/.test(stringValue)) {
    return `"${stringValue.replace(/"/g, '""')}"`;
  }
  return stringValue;
}

File: Java/java_ast/src/main/java/com/ast_tool/output/CsvOutput.java
package com.ast_tool.output;

import com.ast_tool.AppConfig;
import com.ast_tool.model.MethodInfo;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.stream.Collectors;

/**
 * CsvOutput
 * TS版 csv_output.ts の Java移植版。
 * ソースファイルレベルCSV・メソッドレベルCSVの2種類のみを出力する。
 * JSON出力は行わない（JSON全廃に伴い、JsonOutput.java相当のクラスは作成していない）。
 *
 * BOM無しUTF-8で固定出力する（Excelでの文字化け防止。TS版のwriteFile()と同じ対応）。
 * v1のJavaで発生した「config.jsonのencoding設定を無視してMS932固定になっていた」不具合の
 * 反省を踏まえ、v2では設定項目自体を設けず、実装を1本化した。
 */
public class CsvOutput {

    public void writeSourceFileCsv(List<MethodInfo.SourceFileRow> rows, AppConfig config) throws IOException {
        String headerLine = String.join(",", escapeAll(MethodInfo.SOURCE_FILE_ROW_HEADER));

        List<String> lines = rows.stream().map(r -> String.join(",",
                escapeCsv(r.appName),
                escapeCsv(r.fileName),
                escapeCsv(r.directoryPath),
                escapeCsv(r.className),
                escapeCsv(r.importList),
                escapeCsv(String.valueOf(r.lineCount)),
                escapeCsv(String.valueOf(r.methodCount)),
                escapeCsv(String.valueOf(r.variableCount)),
                escapeCsv(String.valueOf(r.constantCount)))).collect(Collectors.toList());

        writeFile(config.getSourceFileCsvFilename(), headerLine + "\n" + String.join("\n", lines), config);
    }

    public void writeMethodCsv(List<MethodInfo.MethodRow> rows, AppConfig config) throws IOException {
        List<String> header = MethodInfo.buildMethodRowHeader();
        String headerLine = String.join(",", escapeAll(header.toArray(new String[0])));

        List<String> lines = rows.stream().map(r -> {
            StringBuilder sb = new StringBuilder();
            sb.append(escapeCsv(r.filePath)).append(",");
            sb.append(escapeCsv(r.className)).append(",");
            sb.append(escapeCsv(r.methodName));
            for (int i = 0; i < MethodInfo.MAX_NEST_DEPTH; i++) {
                sb.append(",").append(escapeCsv(String.valueOf(r.processCoords[i])));
            }
            sb.append(",").append(escapeCsv(r.processContent));
            sb.append(",").append(escapeCsv(r.role));
            sb.append(",").append(escapeCsv(r.returnType));
            sb.append(",").append(escapeCsv(r.methodType));
            sb.append(",").append(escapeCsv(r.accessModifier));
            for (int i = 0; i < MethodInfo.MAX_ARG_COLUMNS; i++) {
                String arg = (i < r.args.length && r.args[i] != null) ? r.args[i] : MethodInfo.HYPHEN;
                sb.append(",").append(escapeCsv(arg));
            }
            return sb.toString();
        }).collect(Collectors.toList());

        writeFile(config.getMethodCsvFilename(), headerLine + "\n" + String.join("\n", lines), config);
    }

    // 【新規】フィールドレベルCSV（3つ目のCSV）の出力
    public void writeFieldCsv(List<MethodInfo.FieldRow> rows, AppConfig config) throws IOException {
        String headerLine = String.join(",", escapeAll(MethodInfo.FIELD_ROW_HEADER));

        List<String> lines = rows.stream().map(r -> String.join(",",
                escapeCsv(r.filePath),
                escapeCsv(r.className),
                escapeCsv(r.methodName == null ? "" : r.methodName),
                escapeCsv(r.fieldKind),
                escapeCsv(r.fieldName),
                escapeCsv(r.fieldType),
                escapeCsv(r.isFinal ? "1" : "0"),
                escapeCsv(String.valueOf(r.validationMin)),
                escapeCsv(String.valueOf(r.validationMax)),
                escapeCsv(String.valueOf(r.nullable)),
                escapeCsv(r.rawAnnotations))).collect(Collectors.toList());

        writeFile(config.getFieldCsvFilename(), headerLine + "\n" + String.join("\n", lines), config);
    }

    private void writeFile(String filename, String content, AppConfig config) throws IOException {
        Path outputDir = Paths.get(config.getOutputDir());
        Files.createDirectories(outputDir);
        Path outputPath = outputDir.resolve(filename);

        try (BufferedWriter writer = Files.newBufferedWriter(outputPath, StandardCharsets.UTF_8)) {
            writer.write(content);
        }
        long lineCount = content.chars().filter(c -> c == '\n').count() + 1;
        System.out.println(filename + ": " + lineCount + "行");
    }

    // escapeCsv: ダブルクォートは全置換（tsjs側で修正済みの挙動を踏襲）
    public static String escapeCsv(String value) {
        if (value == null)
            return "";
        if (value.contains("\"") || value.contains(",") || value.contains("\n") || value.contains("\r")) {
            return "\"" + value.replace("\"", "\"\"") + "\"";
        }
        return value;
    }

    private String[] escapeAll(String[] values) {
        String[] result = new String[values.length];
        for (int i = 0; i < values.length; i++) {
            result[i] = escapeCsv(values[i]);
        }
        return result;
    }
}
